### DISCUSSION TITLE
Add time varying reference tracking support to the objective function calculation

### DISCUSSION THREAD

**Discussion by brilianputraa** [do-mpc/do-mpc#309] [Q&A] [answered]
I wanna use do-mpc for non static/ time varying reference or set point so the optimizer can calculate the cost for future trajectory as well. However I found that I cannot do that due to the size for the constraint term size on the **controller.MPC.set_objective** should be 1x1 or scalar. 
![image](https://user-images.githubusercontent.com/18501809/193647569-40101e2d-6503-4415-88f9-378d9ebbb4b0.png)

In my opinion, it will be really great if we can use time varying set point trajectory path until prediction horizon for the cost function calculation.

**Comment by 4flixt** [comment, marked as answer]
Dear @brilianputraa,

thank you for your question. It is common to have a setpoint for each step of the horizon but you will typically optimize over the weighted sum of these objectives. There is some great literature, e.g. [MPC: Theory, Computation, and Design](https://sites.engineering.ucsb.edu/~jbraw/mpc/) that might help you.

Best,
4flixt

**Reply by brilianputraa** [reply]
Okay thank you for the explanation, but my question is how can we use the future set point if it is different from the present set point?, let's say I wanna control a car, as we know its trajectory at time step t and time step t+1 will be different

**Reply by maqifrnswa** [reply]
See: https://github.com/do-mpc/do-mpc/discussions/269

Set the mterm/lterm to a tvp, then update the tvp with the trajectories at time step t and t+1

**Reply by brilianputraa** [reply]
Thanks!
